Create MissingExactTemplate exception with separate template

vinistock opened the pull request:
This PR is supposed to address issue #29280.

Create new exception class MissingExactTemplate with it's own HTML template and raise it in the implicit_render class.

This is my first contribution, so please let me know if I'm missing something or if I'm breaking any code style standards.

Commit:
Create MissingExactTemplate exception with separate template

matthewd commented on actionpack/lib/action_controller/metal/exceptions.rb:
```diff
@@ -51,4 +51,7 @@ class UnknownHttpMethod < ActionControllerError #:nodoc:
   class UnknownFormat < ActionControllerError #:nodoc:
   end
+
+  class MissingExactTemplate < ActionControllerError #:nodoc:
```
Inherit from UnknownFormat, so the new exception is still rescued by anything that currently matches it

matthewd reviewed (commented):
This is looking about the right shape 👍🏻

Still need this bit:

> Turn the long exception message into a custom HTML template

vinistock commented:
@matthewd Thanks for reviewing the PR. I addressed your comment about the inheritance and fixed the respective tests.

Although I must say I don't quite understand the bit about turning the message into an html template. Is the entire message moving to a template? Will the exception not use it for instantiation anymore? Sorry, not really sure what is expected on this portion.

kaspth commented:
@vinistock try triggering the exception in a newly generated Rails 5.1 app. Notice how the exception is just blurted out as one big chunk of text? We want to do better.

We want the custom template to communicate what the exception *means* better than what the one paragraph exception message can do. To "show" rather than "tell".

The existing exception message can be a spring board, but you can almost do with it as you please, it's not required for instantiation anymore (that's not fully true, but lets get the template going first then we can hone details).

vinistock commented:
@kaspth I've customized the template a bit and changed the message. I also added what the expected path for the template would be, I thought that would make the exception cause more obvious to the user.

Let me know if this is aligned with the expectations and if the exception messages / explanations are as descriptive and concise as they should be.

The layout remains similar to the majority of Rails error pages. If any changes concerning this are wanted, give me a heads up.

matthewd commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -39,18 +39,9 @@ def default_render(*args)
-
-        raise ActionController::UnknownFormat, message
+        message = "#{self.class.name}\##{action_name} is missing a template for request formats: #{request.formats.map(&:to_s).join(',')}"
+        expected_template_path = "app/views/#{self.class.controller_path}/#{action_name}.html.erb"
```
This is an interesting idea, but making a more accurate guess is a lot harder than this, so I think it's best to leave it off for now. (Notably, it has checked a lot more than one filename.)

matthewd commented on actionpack/lib/action_dispatch/middleware/exception_wrapper.rb:
```diff
@@ -23,7 +23,8 @@ class ExceptionWrapper
       "AbstractController::ActionNotFound"  => "unknown_action",
-      "ActionView::Template::Error"         => "template_error"
+      "ActionView::Template::Error"         => "template_error",
+      "ActionController::MissingExactTemplate" => "missing_exact_template"
```
Let's add a trailing comma, so this line doesn't have to change next time we add an entry.

Also, I think we need to add our new exception to `rescue_responses` above as well (pointing to `:not_acceptable`, same as the existing UnknownFormat)

matthewd commented on actionpack/lib/action_dispatch/middleware/templates/rescues/missing_exact_template.html.erb:
```diff
@@ -0,0 +1,23 @@
+<header>
+  <h1>Missing exact template</h1>
```
"exact" sounds a bit odd in a user-facing context here. How about we title it "No template for interactive request"?

matthewd reviewed (commented):
Sorry, I spectacularly dropped the ball on this 😟

I've added a few small notes, but this seems to be just about good to go. I haven't yet actually caused the error locally to see the resulting page, but the HTML sounds good! (If you happen to have it open locally and can add a screenshot, that'd be awesome; if not, that's fine too.)

Sorry again I left this unanswered for so long!

vinistock commented:
@matthewd I've addressed your comments and rebased with the current master. Let me know if any other changes are needed.

Here's the screenshot of a sample application raising the error.

![missing_exact_template](https://user-images.githubusercontent.com/18742907/36040480-b018d374-0dac-11e8-9791-b5ef2dc4ddb0.png)

rafaelfranca commented on actionpack/lib/action_controller/metal/exceptions.rb:
```diff
@@ -50,4 +50,10 @@ class UnknownHttpMethod < ActionControllerError #:nodoc:
   end
+
+  class MissingExactTemplate < UnknownFormat #:nodoc:
+    def initialize(message)
```
Any reason why we need to define an initialize?

rafaelfranca commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -41,18 +41,8 @@ def default_render(*args)
-
-        raise ActionController::UnknownFormat, message
+        message = "#{self.class.name}\##{action_name} is missing a template for request formats: #{request.formats.map(&:to_s).join(',')}"
+        raise ActionController::MissingExactTemplate.new(message)
```
can't we can keep the `raise ActionController::MissingExactTemplate, message` pattern here?

rafaelfranca commented on actionpack/lib/action_dispatch/middleware/exception_wrapper.rb:
```diff
@@ -26,7 +27,8 @@ class ExceptionWrapper
       "ActiveRecord::StatementInvalid"      => "invalid_statement",
-      "ActionView::Template::Error"         => "template_error"
+      "ActionView::Template::Error"         => "template_error",
+      "ActionController::MissingExactTemplate" => "missing_exact_template",
```
can we align all other values or maybe a next PR un-align all the values?

rafaelfranca commented on actionpack/lib/action_dispatch/middleware/templates/rescues/missing_exact_template.html.erb:
```diff
@@ -0,0 +1,23 @@
+    to be rendered for such requests. If that's the case, carry on.
+  </p>
+
+  <%= render template: "rescues/_source" %>
```
Not sure if the source and the trace is going to be useful for this exception. The source is always inside Rails. I'd also remove the request_and_response. This page is more informative than actually an application error.

vinistock commented:
@rafaelfranca Thanks for the review. I believe all your comments make sense and applied all the requested changes. Let me know if there's anything else to be done.

rafaelfranca reviewed (commented):
This looks good to me. Can you squash your commits?

vinistock commented:
@rafaelfranca all done.
